tests/runner.py

```diff
@@ -275,7 +280,104 @@ def create_test_file(name, contents, binary=False):
 test_index = 0
 
 
-class RunnerCore(unittest.TestCase):
+def parameterized(parameters):
+  """
+  Mark a test as parameterized.
+
+  Usage:
+    @parameterized([
+      ('subtest1', (1, 2, 3)),
+      ('subtest2', (4, 5, 6), {'keyword': 'x'}),
+    ])
+    def test_something(self, a, b, c, keyword=None):
+      # actual test body
+
+  This is equivalent to defining two tests:
+
+    def test_something_subtest1(self):
+      self.test_something(1, 2, 3)
+
+    def test_something_subtest2(self):
+      self.test_something(4, 5, 6, keyword='x')
+
+  However, test_something itself is not run as a test.
+  """
+  def decorator(func):
+    func._parameterize = parameters
+    return func
+  return decorator
+
+
+class RunnerMeta(type):
+  @classmethod
+  def make_test(mcs, name, func, suffix, args, kwargs={}):
+    """
+    This is a helper function to create new test functions for each parameterized form.
+
+    :param name: the original name of the function
+    :param func: the original function that we are parameterizing
+    :param suffix: the suffix to append to the name of the function for this parameterization
+    :param args: the positional arguments to pass to the original function for this parameterization
+    :param kwargs: the keyword arguments to pass to the original function for this parameterization
+    :returns: a tuple of (new_function_name, new_function_object)
+    """
+
+    # Create the new test function. It calls the original function with the specified args and kwargs.
+    # We use @functools.wraps to copy over all the function attributes.
+    @wraps(func)
+    def resulting_test(self):
+      return func(self, *args, **kwargs)
+
+    # Add suffix to the function name so that it displays correctly.
+    # We are using `name` instead of `func.__name__` to show the correct name when using decorators that
+    # do not do @functools.wraps.
+    resulting_test.__name__ = '%s_%s' % (name, suffix)
+
+    # On python 3, functions have __qualname__ as well. This is a full dot-separated path to the function.
+    # We add the suffix to it as well.
+    if hasattr(func, '__qualname__'):
+      resulting_test.__qualname__ = '%s_%s' % (func.__qualname__, suffix)
+
+    return resulting_test.__name__, resulting_test
+
+  def __new__(mcs, name, bases, attrs):
+    # A metaclass takes the class name, a list of base classes, and the new class's attributes as a dict.
+    # Then, it returns an instance of the newly created class.
+    # Here, we inspect attrs, the attributes of the class, and create new_attrs, the attributes the
+    # new class will actually be created with.
```

Comment:
It would be good to give more of the motivation here and mention that we're using a metaclass functionality, rather than explaining how metaclasses work generally. Something like
```
# Expand parameterized methods into separate methods in the class object this metaclass is constructing
```